internal/transports/p2p/peer/peer.go

```diff
@@ -0,0 +1,558 @@
+
+	p.log.Info().Msgf("protocol negotiated successfully with peer: %s", p)
+	return nil
+}
+
+// PingHandler is a handler for sending ping messages to peers.
+// Must be run as a goroutine.
+func (p *Peer) pingHandler() {
+	p.wg.Add(1)
+	pingTicker := time.NewTicker(pingInterval)
+	defer pingTicker.Stop()
+
+out:
+	for {
+		select {
+		case <-pingTicker.C:
+			nonce, err := wire.RandomUint64()
+			if err != nil {
+				p.log.Error().Msgf("error generating nonce for ping msg, reason: %v", err)
+				continue
+			}
+			p.log.Info().Msgf("sending ping to peer %s with nonce: %d", p, nonce)
+			p.queueMessage(wire.NewMsgPing(nonce))
+
+		case <-p.quit:
+			break out
+		}
+	}
+
+	p.log.Info().Msgf("ping handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+// MsgHandler is a message handler for incoming messages.
+// Must be run as a goroutine.
+func (p *Peer) readMsgHandler() {
+	p.wg.Add(1)
+
+out:
+	for {
+		select {
+		case <-p.quit:
+			break out
+
+		default:
+			remoteMsg, _, err := wire.ReadMessage(p.conn, p.protocolVersion, p.chainParams.Net)
+			if err != nil {
+				if !p.quitting {
+					p.log.Error().Msgf("cannot read message from peer %s, reason: %v", p, err)
+				}
+				continue
+			}
+
+			switch msg := remoteMsg.(type) {
+			case *wire.MsgPing:
+				p.handlePingMsg(msg)
+			case *wire.MsgPong:
+				p.log.Info().Msgf("received pong from peer %s with nonce: %d", p, msg.Nonce)
+			case *wire.MsgHeaders:
+				p.handleHeadersMsg(msg)
+			case *wire.MsgInv:
+				p.handleInvMsg(msg)
+			default:
+				p.log.Info().Msgf("received msg of type: %T", msg)
+			}
+		}
+	}
+
+	p.log.Info().Msgf("read msg handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+func (p *Peer) writeMessage(msg wire.Message) error {
+	err := wire.WriteMessage(p.conn, msg, p.protocolVersion, p.chainParams.Net)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *Peer) writeRejectMessage(msg wire.Message, reason string) {
+	rejectMsg := wire.NewMsgReject(msg.Command(), wire.RejectObsolete, reason)
+	err := p.writeMessage(rejectMsg)
+	if err != nil {
+		p.log.Error().Msgf("could not write reject message to peer, reason: %v", err)
+	}
+}
+
+func (p *Peer) queueMessage(msg wire.Message) {
+	// running in goroutine here, because writing
+	// to channel is blocking until read
+	go func() {
+		p.msgChan <- msg
+	}()
+}
+
+// writeMsgHandler serves as a queue for writing messages to peers,
+// must be run as a goroutine.
+func (p *Peer) writeMsgHandler() {
+	p.wg.Add(1)
+
+out:
+	for {
+		select {
+		case msg := <-p.msgChan:
+			err := p.writeMessage(msg)
+			if err != nil {
+				p.log.Error().Msgf("error writing msg %T to peer %s", msg, p)
+			}
+		case <-p.quit:
+			break out
+		}
+	}
+
+cleanup:
+	for {
+		select {
+		case <-p.msgChan:
+		default:
+			break cleanup
+		}
+	}
+
+	p.log.Info().Msgf("write msg handler shutdown for peer %s", p)
+	p.wg.Done()
+}
+
+func (p *Peer) writeOurVersionMsg() error {
+	n, err := rand.Int(rand.Reader, big.NewInt(9223372036854775807))
```

Comment:
It looks like a magic number here. Consider making some contant ot add a comment why this actual number.

Reply:
You're right, removed this section completely. I used the much simpler function from `math.rand` `rand.Uint64` for this task

Reply:
Update, the use of `math/rand` created a security comment from golang linter and suggested to use `crypto/rand`, so I did and explained it better, as you suggested